src/components/action-popover/action-popover-item/action-popover-item.component.tsx

```diff
@@ -124,42 +133,96 @@ export const ActionPopoverItem = ({
   );
 
-  const { setOpenPopover, isOpenPopover, focusButton } = context;
+  const {
+    setOpenPopover,
+    isOpenPopover,
+    focusButton,
+    submenuPosition,
+  } = context;
   const isHref = !!href;
   const [containerPosition, setContainerPosition] = useState<
     ContainerPosition | undefined
   >(undefined);
   const [guid] = useState(createGuid());
   const [isOpen, setOpen] = useState(false);
   const [focusIndex, setFocusIndex] = useState<number>(0);
-  const [isLeftAligned, setIsLeftAligned] = useState(true);
 
   const submenuRef = useRef<HTMLDivElement>(null);
   const ref = useRef<HTMLButtonElement>(null);
   const mouseEnterTimer = useRef<NodeJS.Timeout | null>(null);
   const mouseLeaveTimer = useRef<NodeJS.Timeout | null>(null);
+  const hasSubmenuAligned = useRef(false);
 
   useEffect(() => {
     if (!isOpenPopover) {
       setOpen(false);
     }
   }, [isOpenPopover]);
 
+  useEffect(() => {
+    if (icon) {
+      setChildHasIcon?.(true);
+    }
+    if (submenu) {
+      setChildHasSubmenu?.(true);
+    }
+  }, [icon, setChildHasSubmenu, setChildHasIcon, submenu]);
+
   const alignSubmenu = useCallback(() => {
-    if (checkRef(ref) && checkRef(submenuRef) && submenu) {
-      const align = leftAlignSubmenu(ref, submenuRef);
-      setIsLeftAligned(align);
-      setContainerPosition(getContainerPosition(ref, submenuRef, placement));
+    const checkCalculatedSubmenuPosition = calculateSubmenuPosition(
+      ref,
+      submenuRef,
+      submenuPosition,
+      currentSubmenuPosition
+    );
+
+    setCurrentSubmenuPosition?.(checkCalculatedSubmenuPosition);
+
+    return checkRef(ref) && checkRef(submenuRef) && submenu;
+  }, [
+    submenu,
+    setCurrentSubmenuPosition,
+    submenuPosition,
+    currentSubmenuPosition,
+  ]);
+
+  useEffect(() => {
+    function getContainerPosition(
+      itemRef: React.RefObject<HTMLElement>,
+      containerSubmenuRef: React.RefObject<HTMLElement>,
+      containerPlacement: "bottom" | "top"
+    ): ContainerPosition | undefined {
+      /* istanbul ignore if */
+      if (!itemRef.current || !containerSubmenuRef.current) return undefined;
+
+      const { offsetWidth: submenuWidth } = containerSubmenuRef.current;
+
+      const leftAlignedSubmenu = currentSubmenuPosition === "left";
+      const leftValue = leftAlignedSubmenu ? -submenuWidth : "auto";
+      const rightValue = leftAlignedSubmenu ? "auto" : -submenuWidth;
+      const yPositionName = containerPlacement === "top" ? "bottom" : "top";
+
+      return {
+        left: leftValue,
+        [yPositionName]: "calc(-1 * var(--spacing100))",
+        right: rightValue,
+      };
     }
-  }, [submenu, placement]);
+    setContainerPosition(getContainerPosition(ref, submenuRef, placement));
+  }, [submenu, placement, currentSubmenuPosition]);
 
   useEffect(() => {
-    alignSubmenu();
+    if (submenu && !hasSubmenuAligned.current) {
+      alignSubmenu();
+      hasSubmenuAligned.current = true;
+    }
+  });
```

Comment:
```suggestion
  }, [alignSubmenu, submenu]);
```

**nitpick/opinion**: I don't think it's all that important, since by design the `alignSubmenu()` can only be run once by this, but I'd prefer not to run the effect itself on every render. (Performance-wise I don't actually know which is better, not running the `if` check after every render should save time but adding the dependencies makes React itself check if they've changed. I doubt it matters noticeable either way - I just prefer to include effect dependencies. Happy to leave this up to you though.)